algorithms/src/snark/groth16/r1cs_to_qap.rs

```diff
@@ -66,33 +71,20 @@ impl R1CStoQAP {
         let mut b = vec![E::Fr::zero(); qap_num_variables + 1];
         let mut c = vec![E::Fr::zero(); qap_num_variables + 1];
 
-        for i in 0..assembly.num_public_variables {
-            a[i] = u[assembly.num_constraints() + i];
-        }
+        a[..assembly.num_public_variables]
+            .copy_from_slice(&u[assembly.num_constraints()..][..assembly.num_public_variables]);
 
         for (i, x) in u.iter().enumerate().take(assembly.num_constraints()) {
```

Comment:
Maybe we can try additionally parallelizing this loop?

Reply:
I did consider it, but due to the way `a`, `b` and `c` are mutated and accessed I wasn't able to do in an idiomatic way, and the was I _was_ able to do it didn't provide performance benefits.